zds/search2/models.py

```diff
@@ -0,0 +1,283 @@
+from elasticsearch_dsl.connections import connections
+
+
+class ESIndexableMixin(object):
+    """Mixin for indexable objects.
+
+    Define a number of different functions that can be overridden to tune the behavior of indexing into elasticsearch.
+
+    You (may) need to override :
+
+    - ``get_indexable()`` ;
+    - ``get_mapping()`` (not mandatory, but otherwise, ES will choose the mapping by itself) ;
+    - ``get_document()`` (not mandatory, but may be useful if data differ from mapping or extra stuffs need to be done).
+
+    You also need to maintain ``es_id`` and ``es_already_indexed`` for bulk indexing/updating (if any).
+    """
+
+    es_already_indexed = False
+    es_id = ''
+
+    @classmethod
+    def get_es_content_type(cls):
+        """value of the ``_type`` field in the index"""
+        content_type = cls.__name__.lower()
+
+        # fetch parents
+        for base in cls.__bases__:
+            if issubclass(base, ESIndexableMixin) and base != ESDjangoIndexableMixin:
+                content_type = base.__name__.lower() + '_' + content_type
+
+        return content_type
+
+    @classmethod
+    def get_es_mapping(self):
+        """Setup mapping (data scheme).
+
+        .. information::
+            You will probably want to change the analyzer and boost value.
+            Also consider the ``index='not_analyzed'`` option to improve performances.
+
+        See https://elasticsearch-dsl.readthedocs.io/en/latest/persistence.html#mappings
+
+        .. attention::
+            You *may* want to override this method (otherwise ES choose the mapping by itself).
+
+        :return: mapping object
+        :rtype: elasticsearch_dsl.Mapping
+        """
+
+        m = Mapping(self.get_es_content_type())
+        return m
+
+    @classmethod
+    def get_es_indexable(cls, force_reindexing=False):
+        """Get a list of object to be indexed. Thought this method, you may limit the reindexing.
+
+        .. attention::
+            You need to override this method (otherwise nothing will be indexed).
+
+        :param force_reindexing: force to return all objects, even if they may be already indexed.
+        :type force_reindexing: bool
+        :return: list of object to be indexed
+        :rtype: list
+        """
+
+        return []
+
+    def get_es_document_source(self, exclude_field=None):
+        """Create a document from the variable of the class, based on the mapping.
+
+        .. attention::
+            You may need to override this method if the data differ from the mapping for some reason.
+
+        :param exclude_field: exclude some field from the default method
+        :type exclude_field: list
+        :return: document
+        :rtype: dict
+        """
+
+        cls = self.__class__
+        fields = list(cls.get_es_mapping().properties.properties.to_dict().keys())
+
+        data = {}
+
+        for field in fields:
+            if exclude_field and field in exclude_field:
```

Comment:
ah, et très objectivement, je le comprend comme `exclude a field` et pas comme `a field is exluded`. Mais chui pas bien dans ma tête :p